collector-bitmovin-player/src/main/java/com/bitmovin/analytics/bitmovin/player/BitmovinSdkAdAdapter.kt

```diff
@@ -0,0 +1,101 @@
+package com.bitmovin.analytics.bitmovin.player
+
+import com.bitmovin.analytics.BitmovinAdAnalytics
+import com.bitmovin.analytics.adapters.AdAdapter
+import com.bitmovin.analytics.bitmovin.player.utils.AdBreakFactory
+import com.bitmovin.analytics.bitmovin.player.utils.AdFactory
+import com.bitmovin.analytics.bitmovin.player.utils.AdQuartileFactory
+import com.bitmovin.analytics.data.AdModuleInformation
+import com.bitmovin.player.BitmovinPlayer
+import com.bitmovin.player.api.event.listener.*
+
+/**
+ * An adapter that maps the Ad Events to the BitmovinAdAnalytics class
+ */
+class BitmovinSdkAdAdapter(val bitmovinPlayer: BitmovinPlayer, val adAnalytics: BitmovinAdAnalytics) : AdAdapter {
+
+    private val adFactory: AdFactory = AdFactory()
+    private val adBreakFactory: AdBreakFactory = AdBreakFactory()
+    private val adQuartileFactory: AdQuartileFactory = AdQuartileFactory()
+
+
+    private val onAdStartedListener = OnAdStartedListener {
+        if(it.ad != null){
+            adAnalytics.onAdStarted(adFactory.FromPlayerAd(it.ad!!))
+        }
+    }
+
+    private val onAdFinishedListener = OnAdFinishedListener {
+        adAnalytics.onAdFinished()
+    }
+
+    private val onAdBreakStartedListener = OnAdBreakStartedListener {
+        if (it.adBreak != null){
+            adAnalytics.onAdBreakStarted(adBreakFactory.FromPlayerImaAdBreak(it.adBreak!! as com.bitmovin.player.model.advertising.ima.ImaAdBreak))
+        }
+    }
+
+    private val onAdBreakFinishedListener = OnAdBreakFinishedListener {
+        adAnalytics.onAdBreakFinished()
+    }
+
+    private val onAdClickedListener = OnAdClickedListener {
+        adAnalytics.onAdClicked(it.clickThroughUrl)
+    }
+
+    private val onAdErrorListener = OnAdErrorListener {
+        if(it.adConfiguration != null && it.adConfiguration!! is com.bitmovin.player.model.advertising.ima.ImaAdBreak)
```

Comment:
Same here - early return and better type cast clarifies the code

Reply:
implemented early return